src/vs/workbench/services/editor/browser/editorService.ts

```diff
@@ -783,9 +783,9 @@ export class EditorService extends Disposable implements EditorServiceImpl {
 	//#region invokeWithinEditorContext()
 
 	invokeWithinEditorContext<T>(fn: (accessor: ServicesAccessor) => T): T {
-		const activeTextEditorControl = this.activeTextEditorControl;
-		if (isCodeEditor(activeTextEditorControl)) {
-			return activeTextEditorControl.invokeWithinContext(fn);
+		const result = this.activeEditorPane?.invokeWithinContext(fn);
```

Comment:
We have to preserve the semantics of the code how it used to be: check the implementation of `activeTextEditorControl` around `isCompositeEditor`:

```ts
get activeTextEditorControl(): ICodeEditor | IDiffEditor | undefined {
	const activeEditorPane = this.activeEditorPane;
	if (activeEditorPane) {
		const activeControl = activeEditorPane.getControl();
		if (isCodeEditor(activeControl) || isDiffEditor(activeControl)) {
			return activeControl;
		}
		if (isCompositeEditor(activeControl) && isCodeEditor(activeControl.activeCodeEditor)) {
			return activeControl.activeCodeEditor;
		}
	}

	return undefined;
}
```

I think @jrieken added code specifically to support notebooks here. Are you planning to implement this method properly in the notebook editor then to preserve the semantics?

Reply:
Ah, I think I didn't fully understand the composite editor concept - I think that if TitleControl was aware of composite editors, then it could get the ContextKeyService out of the editor in the notebook and we wouldn't have needed any of these changes.

So I guess the question is should this work by default for all composite editors or should every composite editor have to implement this itself.

Reply:
I suggest you coordinate with Jo on this topic, he added the concept.

Reply:
I think that we can either

- Allow any editor to implement/override invokeWithinContext
  - And in this option, need to decide whether it works by default for composite editors or if they still have to implement it (i don't know what other composite editors there are)
- Or leave it as a back door only for text editors, making TitleControl aware of composite editors so that it works for notebooks. In this case I guess we don't have the option to override it for custom/other editors

Reply:
The composite editor is simply a list of code editor of which one might be active. Having the code above ensures that the command palette works for the correct cell editor, but there aren't so many users of `invokeWithEditorContext` so this could be easily changed. 

On a related note: it seems that all users of `invokeWithinEditorContext` are only interested in the "relevant" context key service. So would it maybe be better to simply expose that and leave `ICodeEditorWidget#invokeWith" as an implementation detail?

Reply:
> it seems that all users of invokeWithinEditorContext are only interested in the "relevant" context key service

@bpasero what do you think about this, what else do you expect this to be used for? As-is, it opens up editor internals much more than necessary.

Reply:
Yeah worth a shot. I do not have any stakes in `invokeWithinEditorContext` so any solution is fine with me that does not break existing behaviour.